File: src/saas/saas/static/js/helpers/inventory/items.js
'use strict';

class InventoryItem {

    static find(clientId = '', filter = '', numItems = 20, pageNum = 1) {
        return fetch('/api/inventory/items/filter', {
            method: 'POST',
            credentials: 'same-origin',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                clientId: clientId,
                filter: filter,
                numItems: numItems,
                pageNum: pageNum
            })
        })
        .then((r) => r.json());
    }

    static get(clientId = '', itemId = '') {

    }

    static add(clientId = '', item = {}) {
        return fetch('/api/inventory/items/add', {
            method: 'POST',
            credentials: 'same-origin',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                clientId: clientId,
                name: item.name,
                description: item.description,
                sku: item.sku,
                upc: item.upc,
                make: item.make,
                brand: item.brand,
                model: item.model,
                version: item.version,
                length: item.length,
                length_unit_id: item.length_unit_id,
                width: item.width,
                width_unit_id: item.width_unit_id,
                height: item.height,
                height_unit_id: item.height_unit_id,
                weight: item.weight,
                weight_unit_id: item.weight_unit_id,
                perishable: item.perishable,
                hazardous: item.hazardous
            })
        })
        .then((r) => r.json());
    }

    static substitutes(clientId = '', itemId = '', substituteItemId = '') {
        return fetch('/api/inventory/items/substitutes', {
            method: 'POST',
            credentials: 'same-origin',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                clientId: clientId,
                itemId: itemId,
                substituteItemId: substituteItemId
            })
        })
        .then((r) => r.json());
    }
}
export { InventoryItem };

File: src/saas/saas/app/modules/inventory/stores/items.py
import logging
log = logging.getLogger(__name__)

from saas.app.core.services.connection import ConnectionManager
from saas.app.core.stores.base import BaseStore

from uuid import UUID


class ItemsStore(BaseStore):

    def __init__(self, manager: ConnectionManager, name: str):
        super(ItemsStore, self).__init__(manager, name)

    def add(self, item: dict):
        '''add an inventory item
        '''
        try:
            [(item_id, )] = super(ItemsStore, self).runProcTransactional('inventory.item_add', [
                item['clientId'],
                item['name'],
                item['description'],
                item['make'] if 'make' in item else '',
                item['brand'] if 'brand' in item else '',
                item['model'] if 'model' in item else '',
                item['version'] if 'version' in item else '',
                item['sku'] if 'sku' in item else '',
                item['upc'] if 'upc' in item else '',
                item['length'] if 'length' in item else 0,
                item['lengthUnitId'] if 'lengthUnitId' in item else 1,
                item['width'] if 'width' in item else 0,
                item['widthUnitId'] if 'widthUnitId' in item else 1,
                item['height'] if 'height' in item else 0,
                item['heightUnitId'] if 'heightUnitId' in item else 1,
                item['weight'] if 'weight' in item else 0,
                item['weightUnitId'] if 'weightUnitId' in item else 1,
                item['perishable'] if 'perishable' in item else False,
                item['hazardous'] if 'hazardous' in item else False
            ])
            return item_id
        except Exception as e:
            log.error(e)
            raise Exception('Unable to add inventory item')

    def update(self, item: dict):
        '''update an inventory item
        '''
        try:
            [(item_id, )] = super(ItemsStore, self).runProcTransactional('inventory.item_update', [
                item['clientId'],
                item['itemId'],
                item['name'],
                item['description'],
                item['make'] if 'make' in item else '',
                item['brand'] if 'brand' in item else '',
                item['model'] if 'model' in item else '',
                item['version'] if 'version' in item else '',
                item['sku'] if 'sku' in item else '',
                item['upc'] if 'upc' in item else '',
                item['length'] if 'length' in item else 0,
                item['lengthUnitId'] if 'lengthUnitId' in item else 1,
                item['width'] if 'width' in item else 0,
                item['widthUnitId'] if 'widthUnitId' in item else 1,
                item['height'] if 'height' in item else 0,
                item['heightUnitId'] if 'heightUnitId' in item else 1,
                item['weight'] if 'weight' in item else 0,
                item['weightUnitId'] if 'weightUnitId' in item else 1,
                item['perishable'] if 'perishable' in item else False,
                item['hazardous'] if 'hazardous' in item else False
            ])
            return item_id
        except Exception as e:
            log.error(e)
            raise Exception('Unable to update inventory item')

    def filterItems(self, clientId: UUID, filter: str, numItems: int, pageNum: int):
        '''retrieve all matching inventory items
        '''
        try:
            result = super(ItemsStore, self).runProc('inventory.items_filter', [clientId, f'%{filter}%', numItems, pageNum])
            return result
        except Exception as e:
            log.error(e)
            raise Exception('An error occured while retrieving inventory items')

    def substitutes(self, clientId: UUID, itemId: UUID):
        try:
            result = super(ItemsStore, self).runProc('inventory.item_substitutes', [clientId, itemId])
            return result
        except Exception as e:
            log.error(e)
            raise Exception('Unable to retrieve Inventory Item Substitutes')

    def addSubstitute(self, clientId: UUID, itemId: UUID, substituteItemId: UUID):
        try:
            super(ItemsStore, self).runProcTransactional('inventory.items_add_substitute', [
                clientId, itemId, substituteItemId
            ])
        except Exception as e:
            log.error(e)
            raise Exception('An error occured while adding a substitute')